pkg/internal/server/grpc/get_metadata_delta.go

```diff
@@ -135,22 +137,76 @@ func (s *Server) convertToCSIGetMetadataDeltaRequest(ctx context.Context, req *a
 }
 
 func (s *Server) streamGetMetadataDeltaResponse(ctx context.Context, clientStream api.SnapshotMetadata_GetMetadataDeltaServer, csiStream csi.SnapshotMetadata_GetMetadataDeltaClient) error {
+	var (
+		blockMetadataType   api.BlockMetadataType
+		lastByteOffset      int64
+		lastSize            int64
+		logger              = klog.FromContext(ctx)
+		numBlockMetadata    int
+		responseNum         int
+		volumeCapacityBytes int64
+	)
+
 	for {
 		csiResp, err := csiStream.Recv()
 		if err == io.EOF {
-			klog.FromContext(ctx).V(HandlerTraceLogLevel).Info("stream EOF")
+			logger.V(HandlerTraceLogLevel).WithValues(
+				"blockMetadataType", blockMetadataType.String(),
+				"lastByteOffset", lastByteOffset,
+				"lastSize", lastSize,
+				"lastResponseNum", responseNum,
+				"volumeCapacityBytes", volumeCapacityBytes,
+			).Info("stream EOF")
 			return nil
 		}
 
-		//TODO: stream logging with progress
-
 		if err != nil {
+			logger.WithValues(
+				"blockMetadataType", blockMetadataType.String(),
+				"lastByteOffset", lastByteOffset,
+				"lastSize", lastSize,
+				"lastResponseNum", responseNum,
+				"volumeCapacityBytes", volumeCapacityBytes,
+			).Error(err, msgInternalFailedCSIDriverResponse)
 			return s.statusPassOrWrapError(err, codes.Internal, msgInternalFailedCSIDriverResponseFmt, err)
 		}
 
+		responseNum++
+
 		clientResp := s.convertToGetMetadataDeltaResponse(csiResp)
+		blockMetadataType = clientResp.BlockMetadataType
+		volumeCapacityBytes = clientResp.VolumeCapacityBytes
+		numBlockMetadata = len(clientResp.BlockMetadata) - 1
+		if numBlockMetadata >= 0 {
+			lastByteOffset = clientResp.BlockMetadata[numBlockMetadata].ByteOffset
+			lastSize = clientResp.BlockMetadata[numBlockMetadata].SizeBytes
+		}
+
+		if logger.V(HandlerTraceLogLevel).Enabled() {
+			var b strings.Builder
+			b.WriteString("[")
+			for _, bmd := range clientResp.BlockMetadata {
+				b.WriteString(fmt.Sprintf("{%d,%d}", bmd.ByteOffset, bmd.SizeBytes))
+			}
+			b.WriteString("]")
```

Comment:
Can we not use json encoding?

Reply:
JSON encoding is very verbose and would repeat the `byte_offset` and `size_bytes` labels for each entry. This is more compact and expresses the same data.